Cleanup Extern types

webmaster128 opened the pull request.

webmaster128 commented on packages/std/src/traits.rs:
```diff
@@ -3,16 +3,16 @@ use std::vec::Vec;
+/// Designed to allow easy dependency injection at runtime.
+/// This cannot be copied or cloned since it would behave differently
+/// for mock storages and a bridge storage in the VM.
 pub struct Extern<S: Storage, A: Api> {
```
hmm, how was it even possible to derive `Clone` before, when `Storage` is not clonable?

webmaster128 replied:
This is a rabbit hole of Rust compiler specifics. There are a bunch of open and closed Rust issues around that topic, which I hardly understand. It is very well possible, that this is a bug or known limitation in the compiler or the standard. I better give up trying to understand this.

webmaster128 commented:
- go-cosmwasm builds sucessfully with this change
- cosmwasm v0.6.4 (before the big recycle refactorings) also builds successfully with non-clonable `Extern`
- I reverted changes of `ExternalStorage`, which now required an explicit `.clone()` instead of implicit copy.

I'm pretty confident this makes sense, but also those are breaking changes and we can do them in 0.8 instead of here.

ethanfrey commented:
> I reverted changes of ExternalStorage, which now required an explicit .clone() instead of implicit copy.

Can you send a link to the code where this is needed?

webmaster128 commented:
> Can you send a link to the code where this is needed?

`ExternalStorage` was `Clone`able all the time. I added `Copy` because it it stateless, but this was one step in the wrong direction, which is what I reverted.

I doubt that any reasonable code needs to clone `ExternalStorage`, which is why I added the warning comment in #172. https://github.com/CosmWasm/cosmwasm/pull/171/commits/ca21499ca9fc909079c1ed3a7d08275c463d7261 goes one step further and removes the cloning ability from `ExternalStorage`, which seems to work well for both this repo as and go-cosmwasm.

webmaster128 commented:
I ran another round of tests using the erc20 contract with

```toml
[patch.crates-io]
cosmwasm = { path = "../../cosmwasm/packages/std", version = "0.7.1" }
cosmwasm-vm = { path = "../../cosmwasm/packages/vm", version = "0.7.1", default-features = false }
```

which shows no issues.
